pyFAI/ext/CSC_common.pxi

```diff
@@ -0,0 +1,308 @@
+            bint do_azimuthal_variance = error_model is ErrorModel.AZIMUTHAL
+            bint do_variance, is_valid, do_dark, do_flat, do_polarization, check_dummy, do_solidangle, do_absorption
+            preproc_t value
+            acc_t delta1, delta2, b, omega_A, omega_B, omega2_A, omega2_B
+
+
+        
+        assert weights.size == self.input_size, "weights size"
+        empty = dummy if dummy is not None else self.empty
+
+
+        if (dummy is not None) and (delta_dummy is not None):
+            check_dummy = True
+            cdummy = <data_t> float(dummy)
+            cddummy = <data_t> float(delta_dummy)
+        elif (dummy is not None):
+            check_dummy = True
+            cdummy = <data_t> float(dummy)
+            cddummy = 0.0
+        else:
+            check_dummy = False
+            cdummy = <data_t> float(empty)
+            cddummy = 0.0
+    
+        if variance is not None:
+            assert variance.size == self.input_size, "variance size"
+            do_variance = True
+            cvariance = numpy.ascontiguousarray(variance.ravel(), dtype=data_d)
+        else:
+            do_variance = error_model is not ErrorModel.NO
+    
+        if dark is not None:
+            do_dark = True
+            assert dark.size == self.input_size, "dark current array size"
+            cdark = numpy.ascontiguousarray(dark.ravel(), dtype=data_d)
+        else:
+            do_dark = False
+        if flat is not None:
+            do_flat = True
+            assert flat.size == self.input_size, "flat-field array size"
+            cflat = numpy.ascontiguousarray(flat.ravel(), dtype=data_d)
+        else:
+            do_flat = False
+        if polarization is not None:
+            do_polarization = True
+            assert polarization.size == self.input_size, "polarization array size"
+            cpolarization = numpy.ascontiguousarray(polarization.ravel(), dtype=data_d)
+        else:
+            do_polarization = False
+        if solidangle is not None:
+            do_solidangle = True
+            assert solidangle.size == self.input_size, "Solid angle array size"
+            csolidangle = numpy.ascontiguousarray(solidangle.ravel(), dtype=data_d)
+        else:
+            do_solidangle = False
+        if absorption is not None:
+            do_absorption = True
+            assert absorption.size == self.input_size, "absorption array size"
+            cabsorption = numpy.ascontiguousarray(absorption.ravel(), dtype=data_d)
+        else:
+            do_absorption = False
+    
+        with nogil:
+            #first loop for pixel in input image
+            for idx in range(self._indptr.shape[0]-1):  
+                # skip pixel if masked:
+                start = self._indptr[idx]
+                stop = self._indptr[idx+1]
+                if stop == start: # pixel contributes to no bin
+                    continue
+                is_valid = preproc_value_inplace(&value,
```

Comment:
Does this get inlined?

Reply:
Found in pyFAI/ext/regrid_common.pxi: it should be inlined. But I am surprised that norm is not being cached?
`norm = normalization_factor * flat * polarization * solidangle * absorption`. 

A related question: can the dark subtraction be done on the 1D integrated data instead of the 2D image. e.g:

`signal = sum( pixel - dark ) / sum( normalisation ) = ( sum( pixel ) - sum( dark ) ) / sum( normalisation )`

Like this, it appears it would read 6 x 2D images [ data, dark, flat, polarization, solidangle, absorbtion ] per integration] as well as a mask. For a series of images with the same geometry (etc) then I wonder if numbers can be computed by only reading the data, and then subtracting dynamically masked pixels from cached 1D sums?

Reply:
Several question: yes, up to 6 arrays can be read for a a single integration. 
About the dark, if it is constant, you can correct your data a posteriori. The same is true if all 4 normalization arrays are constant (discarding dynamic masking). 

But if you want to keep dynamic masking (and you probably need it with your Eiger detector :-) the correction is possible but probably longer than the intergration itself.